### ISSUE TITLE
sendDuetGCode failed with code: 405and reason: Method Not Allowed

### ISSUE THREAD

**Issue by keyoss** [stuartofmt/DuetLapse3#4]
RRF 3.3.0 2021-06-15
Duet3 Mini5+ ETH
SBC

/usr/bin/python3 /home/pi/DuetLapse3/DuetLapse3.py -port 8082 -duet 192.168.0.216 -detect layer -camera1 pi -pause yes -movehead 270 270 -basedir /home/pi/TimeLapse

error thrown:

Jan 28 15:43:25 Duet3 python3[1183]: 192.168.0.216 Camera1: capturing frame 0 at layer 1 after layer change
Jan 28 15:43:25 Duet3 python3[1183]: 192.168.0.216 Working directory created at: /home/pi/TimeLapse/duet3/192-168-0-216/1183_Cable_ChainClosings_97_Scaled
Jan 28 15:45:31 Duet3 python3[1183]: 192.168.0.216 Requesting pause via M25
Jan 28 15:45:31 Duet3 python3[1183]: 192.168.0.216 sendDuetGCode failed with code: 405and reason: Method Not Allowed
Jan 28 15:45:42 Duet3 python3[1183]: 192.168.0.216 Loop exceeded: Target was: paused
Jan 28 15:45:42 Duet3 python3[1183]: 192.168.0.216 Camera1: capturing frame 1 at layer 2 after layer change
Jan 28 15:47:12 Duet3 python3[1183]: 192.168.0.216 Requesting pause via M25
Jan 28 15:47:12 Duet3 python3[1183]: 192.168.0.216 sendDuetGCode failed with code: 405and reason: Method Not Allowed
Jan 28 15:47:23 Duet3 python3[1183]: 192.168.0.216 Loop exceeded: Target was: paused
Jan 28 15:47:23 Duet3 python3[1183]: 192.168.0.216 Camera1: capturing frame 2 at layer 3 after layer change

Thank you :)

**Comment by stuartofmt** [comment]
Hi - this was one piece of code that I was unable to test because I do not have a Duet SBC.  Alas - no-one else has tested as it  seems not that many use the head repositioning option  :-).
   
I even have a comment in my code that it could fail and know the likely fix.

Give me a day or so (kinda busy) and I'll work directly with you to prove the fix.

**Comment by stuartofmt** [comment]
@keyoss - have you had a chance to test this ?

**Comment by keyoss** [comment]
I just updated some minutes ago... sry i was in UK for work.

Actually it doesnt take any pictures at all even when i manually snapshot. Logfiles and folders are created, it says X Images taken but the directory is still empty. i need to check whats wrong first.

regards
k0s

**Comment by stuartofmt** [comment]
Anything I can do to help - let me know.
Are you  using the ribbon cable attached pi camera?  What OS version are you using ?  The reason I ask is that Pi Debian Bullseye changed things and I have not been able to test that combination.
In any case - there should be some error messages.
While you are testing - running from a command line will be the most beneficial as you will everything as it happens - as well attaching log files will help.

**Comment by keyoss** [comment]
Hey @stuartofmt ,

its Buster, ribbon connected PiCam.

using : `python3 ./DuetLapse3.py  -duet 192.168.0.216 -seconds 20 -detect none -camera1 pi`

works  - it creates and saves imagesequences and while ctrl-c it convertes by ffmpeg a nice video (mp4).

using `python3 ./DuetLapse3.py  -duet 192.168.0.216 -port 8082 -standby -camera1 pi`

logfile will be created, directory created while hitting "Snapshot" in the webcontrol:

`Process ID:    1124

This instance is using the following options:
python3 ./DuetLapse3.py -duet 192.168.0.216 -port 8082 -standby -camera1 pi

Logs and Videos are located here:
/home/pi/DuetLapse3/duet3/192-168-0-216

Images are located here:
/home/pi/DuetLapse3/duet3/192-168-0-216/1124

The logfile for this instance is:
/home/pi/DuetLapse3/duet3/192-168-0-216/1124_22-02-16T22:07:32.log`

i can hmmer as many times the snapshot button, have an ls -al in the console and nothing is created...
[29195_22-02-16T09˸37˸21.log](https://github.com/stuartofmt/DuetLapse3/files/8083614/29195_22-02-16T09.37.21.log)
[3058_22-02-16T17˸17˸02.log](https://github.com/stuartofmt/DuetLapse3/files/8083615/3058_22-02-16T17.17.02.log)

**Comment by stuartofmt** [comment]
I'll take a closer look tomorrow - but a couple of things to note:
1) Snapshot will not work if there are less than 10 images (error message will be in the log file).
2) If a snapshot is taken - it's likely that another request within 10 seconds will be ignored (need to recheck the code).

Nonetheless - that does not explain the missing video.  I'll do dome tests tomorrow and try to recreate.

**Comment by stuartofmt** [comment]
So I did some tests using the following form (good for testing)

-duet 192.168.0.216 -port 8082 -camera1 pi -seconds 5 -detect none

I noticed that no images were being captured.  THEN I remembered that I had videostream running (in my case it starts at boot).  I stopped videostream and all is well.

The raspberry pi libraries only allow ONE connection to the camera  (first to connect "owns" the camera).  This is in part why I wrote videostream.

In any case - I'm wondering I'm you have something like that going on ?
Does the following command (with DuetLapse NOT running) create an image file?  If not its likely some other process "owns" the camera.

```
raspistill -o ./test_image.jpg

```

In my personal use - I have videostream running all the time and use "-camera other" together with -camparam and -weburl for DuetLapse as that allows me to capture timelapses AND view the camera stream in a browser at the same time.

I'll take a look to see if I can trap that situation and report an error - but in the meantime, I suspect that is the current problem on your machine.

**Comment by keyoss** [comment]
Hey,

indeed my RasPi_Webcam_shizzle blocked it. While i still wonder why i could use the 20secs option and capture images...
Anyway i switched to videostream - (i had to install imutils manually which is not pointed in your procedure btw)

I actually test the SBC with Layerchange detection and it creates now successfully imagesequences - so i geuss the ffmpeg conversion to a video will work too.

In around 2hrs i will check the -movehead at layerchange and report back.

Thanks so far!

**Comment by keyoss** [comment]
Detect Layerchange - Movehead:

`DuetLapse3.service - startDuetLapse3 Service
   Loaded: loaded (/lib/systemd/system/DuetLapse3.service; disabled; vendor preset: enabled)
   Active: active (running) since Fri 2022-02-18 13:24:24 CET; 7min ago
 Main PID: 3518 (python3)
    Tasks: 3 (limit: 3597)
   CGroup: /system.slice/DuetLapse3.service
           └─3518 /usr/bin/python3 /home/pi/DuetLapse3/DuetLapse3.py -port 8082 -duet 192.168.0.216 -detect layer -pause yes -movehead 250 250 -weburl1 http://192.168.0.216:8090/stream -camera1 other "-camparam1=ffmpeg +ffmpegquiet +-y -i+weburl+-vframes 1+fn+debug"

Feb 18 13:27:01 Duet3 python3[3518]: 192.168.0.216 !!!!! http status request received !!!!!
Feb 18 13:28:01 Duet3 python3[3518]: 192.168.0.216 !!!!! http status request received !!!!!
Feb 18 13:29:01 Duet3 python3[3518]: 192.168.0.216 !!!!! http status request received !!!!!
Feb 18 13:29:07 Duet3 python3[3518]: 192.168.0.216 Requesting pause via M25
Feb 18 13:29:18 Duet3 python3[3518]: 192.168.0.216 Loop exceeded: Target was: paused
Feb 18 13:29:18 Duet3 python3[3518]: 192.168.0.216 Camera1: capturing frame 1 at layer 2 after layer change
Feb 18 13:29:20 Duet3 python3[3518]: 192.168.0.216 !!!!! http status request received !!!!!
Feb 18 13:31:13 Duet3 python3[3518]: 192.168.0.216 Requesting pause via M25
Feb 18 13:31:24 Duet3 python3[3518]: 192.168.0.216 Loop exceeded: Target was: paused
Feb 18 13:31:24 Duet3 python3[3518]: 192.168.0.216 Camera1: capturing frame 2 at layer 3 after layer change
`

Capture success, BUT Head will not be moved to desired location

**Comment by keyoss** [comment]
Aye,

the python process with the videostream takes the most cpu usage actually: 170% ~ which is 1+ cores in the end. The amount of ram i dont care its a Raspi4 with 4gb. I used to print several parts right now and didnt have any bad collisions so far. I noticed issues due to bandwith with the raspicam when using the videostream with default options, therefore i used -size 2 as incl BGR3 and ofcourse 180° rotation ;)

havin the head moved of the spot would be a great benefit, especially if i print small parts - looking forward to do more testing and feel free to message me anytime.

thanks!

**Comment by stuartofmt** [comment]
@keyoss 

This is a really quick test hack and does not cover all the possible forms.  If you could try it and tell me if any work.
```
Call completed with response: <Response [200]>
```

Make sure your printer is homed and bed is away from the nozzle.

You'll need to rename the file as <something>.py  it should have the correct ip for your printer at the top.



[test2.txt](https://github.com/stuartofmt/DuetLapse3/files/8099380/test2.txt)

**Comment by keyoss** [comment]
[testcode.log](https://github.com/stuartofmt/DuetLapse3/files/8101942/testcode.log)

jfi - again the same behavior - only the first 2 times it moves the head - other 2 not, and again 2 times - then not.

since the response is always 200 and call completed.

**Comment by stuartofmt** [comment]
Thanks for this.  It is (to me) very strange that there are no errors codes being returned in the http requests - something that I will ask the developer(s) about.  This helps in that we know the movement codes are working and the correct format to use.  The M25 error is now the mystery as it's just a command like the movement commands.

In the attached test program (on my standalone machine) I get the expected responses.  The code first sets a slow movement speed then moves the print head in a square.  It then tries various other things including a M25 midway in the code as well as several M25 commands in different forms.  It finishes with another square with a M226 (embedded pause).

The M25 command do not actually pause in this test because M25 only applies when printing from a file.  During the test - in the DWC console there should be messages **Error: M25: Cannot pause print, because no file is being printed!**    

What I'm hoping to see is if there are M405 (or other) errors being returned as you originally saw.

If you would be so kind as to try this again and send back the log.   If this attempt does not provide some clue as to the solution - I'll escalate in the Duet forum as it could be a bug at that end.


[test4.txt](https://github.com/stuartofmt/DuetLapse3/files/8102851/test4.txt)

**Comment by keyoss** [comment]
Ok will test it in some minutes and provide you a video as well... since there is no square movement at all.. its just moving from 0,0 to the center and thats it.

will check the console as well.

thanks for this!

**Comment by keyoss** [comment]
[testcode2.log](https://github.com/stuartofmt/DuetLapse3/files/8102989/testcode2.log)
 ;) now i see the rectangle you been referring to ;D

**Comment by stuartofmt** [comment]
Great - almost there.  One last test (whenever you are ready).

1) Open DWC in browser.   With no print running - run pause.py.  You should see an error in DWC **Error: M25: Cannot pause print, because no file is being printed!**  Note that pause.py does not try to move the print head.

2) Start a print normally.   Once its printing run pause.py.  It should pause as if you had hit pause in DWC.  You should then be able to resume the print from DWC.  You can repeat this if you like :-) 

[pause.txt](https://github.com/stuartofmt/DuetLapse3/files/8103083/pause.txt)

**Comment by keyoss** [comment]
Without print - just idle - the pause.py will not cause the printer to receive a pause command. Manually sending the M25 in the console will give me the error message. 

During Print - Pause command executed (pause.py) and Head is moved to the origin 0,0.
Resume works - pause.py again and again works.

As mentioned : Only during a running print, not when no print is started.

**Comment by stuartofmt** [comment]
Great - the main thing was that it does pause a print in progress.

**In the next day or two - I'll send a new DuetLapse3.py for you to try at your convenience.**

I'm curious (but not necessarily concerned) at the lack of message when idle.  There is supposed to be no different between API call and DWC.  It's possible there  something strange going on either at the firmware level or at the DWC level.

For reference, I'm on DWC 3.3.0  - my machine specific info is:
![image](https://user-images.githubusercontent.com/23531088/154823658-8c703091-7ad7-45c3-b8b0-027f935450eb.png)

**Comment by keyoss** [comment]
![image](https://user-images.githubusercontent.com/8471059/154823800-97dbff16-ef21-4006-8642-f5acee2d38a5.png)
;)

**Comment by stuartofmt** [comment]
@keyoss 

When you get a chance - please try this version.  I've made a few tweaks, added some additional debugging and incorporated the SBC change - so pauses should now work.

Please note that when Duet is asked to pause - it runs a macro called pause.g.  pause.g may have commands to reposition the head.  If this is the case and you use use -movehead then there will be two separate moves of the head.  Not bad of itself but in many cases will be unnecessary.
 
[DuetLapse3.txt](https://github.com/stuartofmt/DuetLapse3/files/8111294/DuetLapse3.txt)

**Comment by keyoss** [comment]
@stuartofmt 
so far so good, i replaced the pause.g with a blanc one.
Only thing i wonder about is the movement speed while the -movehead operation: 4mm/s !?

regards and big ups!

**Comment by stuartofmt** [comment]
Good to hear.

The speed of -movehead is not being controlled by DuetLapse3.  I suspect its some kind of default if there is a blank pause.g

Attached is my pause.g and resume.g with a suggested change if you want to use -movehead.
Basically just giving it a move speed G0 F6000 without any position and then DuetLapse3 provides the position.   The downside is that a normal pause will just lift the print head (which may be OK).

In any case - you can control the speed with pause.g


[pause-resume macro.txt](https://github.com/stuartofmt/DuetLapse3/files/8117889/pause-resume.macro.txt)

**Comment by stuartofmt** [comment]
@keyoss 

Can you explain 
_Also the images are still taken at different positions while the "movehead operation"_

I noticed a bit of "wiggling" of the printhead position which  looked to be due to some latency between the request to capture and the camera feed.

By "wiggle" I mean  that in the video - the position of the print head seemed to move about a bit but very close to the requested position.   

I increased the time between when the end of the -movehead completed and the image was taken and that seemed to fix it.  On my setup - that solved the "wiggles".

**Comment by stuartofmt** [comment]
@keyoss 

Here is the version that I will publish in the next few days.  I've added another option -rest n  where n is the number of seconds to wait after a pause before capturing an image.  In my tests with -rest 3 there  was no "wobble" of the print head in the video and with -rest 0 it was quite noticeable.  This is really a tuning parameter depending on how DuetLapse3 is used.

[DuetLapse3.txt](https://github.com/stuartofmt/DuetLapse3/files/8121015/DuetLapse3.txt)

**Comment by keyoss** [comment]
Awesome,

i checked with your pause and resume and its the same like mine, i also checked the blanc pause file. still my headmovement is just 4mm/sec and i cant figure out how to increase the 4mm/sec to something decent like 20-XXmm/sec.

i will give the rest option a try in some hours and report back.

regards

**Comment by stuartofmt** [comment]
Is it slow if you just send a pause from DWC? 

DWC just sends an M25 code and that causes pause.g to run.  DuetLapse3 does the same thing - just sends an M25 Code.  Then DuetLapse3 sends a G0 Xx Yx where x and y are the -movehead coordinates.  The G0 should inherit the last G0 Ff **or** (as far as I know)  G1 Ff command (which (in this case) comes from pause.g.

I'll do some experiments - to see if G0 Ff and G1 Ff do the same thing (I believe they should be but maybe not).

If they are different then adding a separate G0 Ff line to pause.g might be all that is needed.

P.S. I used G0 in DuetLapse3 because its intended for non-printing moves.  At least that is my understanding.

**Comment by stuartofmt** [comment]
@keyoss 
The pause request has 10 seconds to achieve a status of pause otherwise it will error.
If the  speed is only  4mm/s then it would error if the head has to travel more than 40mm.   Not far at all !!

Can you confirm how / why you get the 4mm/s number as the speed?

Thx

**Comment by keyoss** [comment]
So when i use the M25 with the generic pause.g the speed is around 72mm/s 

`M83            ; relative extruder moves
G1 E-10 F3600   ; retract 4mm of filament
G91            ; relative positioning
G1 Z5 F360     ; lift Z by 5mm
G90            ; absolute positioning
G1 X0 Y0 F6000 ; go to X=0 Y=0
G0 F6000`

using the duetlapse will result in 4mm/s -.-
i gonna check the duetlapse without defining the -movehead now

**Comment by stuartofmt** [comment]
This is vey strange and is not something I can reproduce on my standalone Duet.

The -movehead option only sends G0 Xx Yx where x and y are the movehead positions.
There is no F component.
What's supposed to happen is the speed of the head movement is determined by the last G0 ... Ff  or G1 ... Ff.

I tested this with all combinations of G code and it does not matter if the speed is set by G0 or G1 or if the moves are made by G0 or G1.

Lets say that you used -movehead 100 100  This would be the equivalent of  (in your case)

~~~
M83                     ; relative extruder moves
G1 E-10 F3600     ; retract 4mm of filament
G91                      ; relative positioning
G1 Z5 F360          ; lift Z by 5mm
G90                      ; absolute positioning
G1 X0 Y0 F6000   ; go to X=0 Y=0 G0 F6000
G0 X100 Y100
~~~
(note your retraction is 10mm not 4mm)

From some forum posts -- G0 can have some form of default MAXIMUM speed but only if Duet is used for Laser or CNC - otherwise it's supposed to work as I described above (which is the way it works on my standalone machine)

So at this point - I'm suspecting a bug in the SBC code. (or you accidentally have your machine set to Laser or CNC -- check to see if you have a M453 anywhere in the config.g file).  I doubt in because this should result in fast moves not slow moves 👎 

Lets - try this:  Here is a version of DuetLapse3 with the G0 Xx Yy replaced by G1 Xx Yx.
Lets see if that solves the problem.  If it does - it can be left as is but I will report a possible bug in the forum.


[DuetLapse3.txt](https://github.com/stuartofmt/DuetLapse3/files/8134680/DuetLapse3.txt)
